src/fft.rg

```diff
@@ -13,74 +13,135 @@
 -- limitations under the License.
 
 --[[--
-Regent FFT library.
+Regent FFT library
 ]]
 
 import "regent"
 
+local format = require("std/format")
 local data = require("common/data")
-local cudahelper = require("regent/cudahelper")
+local gpuhelper = require("regent/gpu/helper")
+local gpu_available = gpuhelper.check_gpu_available()
 
+-- Import C and FFTW APIs
 local c = regentlib.c
 local fftw_c = terralib.includec("fftw3.h")
-terralib.linklibrary("libfftw3.so")
+regentlib.linklibrary("libfftw3.so")
+regentlib.linklibrary("libfftw3f.so")
 
-local use_cuda = cudahelper.check_cuda_available()
+-- Import cuFFT API
 local cufft_c
-if use_cuda then
-  cufft_c = terralib.includec("cufft.h")
-  terralib.linklibrary("libcufft.so")
+local cufft_assert
+if gpu_available then
+  cufft_c = terralib.includec("cufftXt.h")
+  regentlib.linklibrary("libcufft.so")
+
+  terra cufft_assert(result : cufft_c.cufftResult)
+    var status = "UNKNOWN"
+    if result == 0 then
+      status = "CUFFT_SUCCESS"
+    elseif result == 1 then
+      status = "CUFFT_INVALID_PLAN"
+    elseif result == 2 then
+      status = "CUFFT_ALLOC_FAILED"
+    elseif result == 3 then
+      status = "CUFFT_INVALID_TYPE"
+    elseif result == 4 then
+      status = "CUFFT_INVALID_VALUE"
+    elseif result == 5 then
+      status = "CUFFT_INTERNAL_ERROR"
+    elseif result == 6 then
+      status = "CUFFT_EXEC_FAILED"
+    elseif result == 7 then
+      status = "CUFFT_SETUP_FAILED"
+    elseif result == 8 then
+      status = "CUFFT_INVALID_SIZE"
+    elseif result == 9 then
+      status = "CUFFT_UNALIGNED_DATA"
+    elseif result == 10 then
+      status = "CUFFT_INCOMPLETE_PARAMETER_LIST"
+    elseif result == 11 then
+      status = "CUFFT_INVALID_DEVICE"
+    elseif result == 12 then
+      status = "CUFFT_PARSE_ERROR"
+    elseif result == 13 then
+      status = "CUFFT_NO_WORKSPACE"
+    elseif result == 14 then
+      status = "CUFFT_NOT_IMPLEMENTED"
+    elseif result == 15 then
+      status = "CUFFT_LICENSE_ERROR"
+    elseif result == 16 then
+      status = "CUFFT_NOT_SUPPORTED"
+    end
+    regentlib.assert(result == cufft_c.CUFFT_SUCCESS, status)
+  end
 end
 
--- Hack: get defines from fftw3.h
+-- Define constants: get defines from fftw3.h
 fftw_c.FFTW_FORWARD = -1
 fftw_c.FFTW_BACKWARD = 1
-
 fftw_c.FFTW_MEASURE = 0
-fftw_c.FFTW_DESTROY_INPUT = (2 ^ 0)
-fftw_c.FFTW_UNALIGNED = (2 ^ 1)
-fftw_c.FFTW_CONSERVE_MEMORY = (2 ^ 2)
-fftw_c.FFTW_EXHAUSTIVE = (2 ^ 3) -- NO_EXHAUSTIVE is default
-fftw_c.FFTW_PRESERVE_INPUT = (2 ^ 4) -- cancels FFTW_DESTROY_INPUT
-fftw_c.FFTW_PATIENT = (2 ^ 5) -- IMPATIENT is default
 fftw_c.FFTW_ESTIMATE = (2 ^ 6)
-fftw_c.FFTW_WISDOM_ONLY = (2 ^ 21)
 
 local fft = {}
 
-function fft.generate_fft_interface(itype, dtype)
+--- Create the FFT interface.
+-- @param itype Index type of transform (int2d/int2d/int3d/int4d).
```

Comment:
```suggestion
-- @param itype Index type of transform (int1d/int2d/int3d/int4d).
```

Also I thought this was limited to 3D?

Reply:
I indicated int4d here since for a 3D batch transform, the user has to pass in a 4D region. But this will change if we switch to the user passing in a int3d for a 3D batch transform

Reply:
Limited to 3D now that we are only taking N and not N+1 in batched transforms